Mathlib/Data/Greedoid/Predefs.lean

```diff
@@ -0,0 +1,193 @@
+/-
+Copyright (c) 2024 Jihoon Hyun. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jihoon Hyun
+-/
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finset.Card
+
+/-!
+# Exchange Property and Accessible Property of Greedoid
+Greedoid is a set system, i.e., a family of sets, over a finite ground set, which contains
+an empty set and satisfies both exchange and accessible properties.
+If a set system `S` satisfies the exchange property, then there is some element `x ∈ s₂ \ s₁`
+which `s₁ ∪ {x} ∈ S`, for every set `s₁, s₂ ∈ S` satisfying `s₁.card < s₂.card`.
+If a set system `S` satisfies the accessible property, then there is some element `x ∈ s`
+which `s \ {x} ∈ S` for every nonempty set `s ∈ S`.
+These two properties are defined in this file as `ExchangeProperty` and `AccessibleProperty`.
+
+While it is sufficient to define a greedoid using only the definitions of the properties, it turned
+out to be useful to have some theorems equipped with them.
+Therefore this file defines exchange property and accessible property, and hands out some basic
+theorems which come along with it, prior to defining a greedoid.
+-/
+
+namespace Greedoid
+
+open Nat Finset
+
+/-- The exchange property of greedoid.
+    Note that the exchange property also hold for matroids. -/
+def ExchangeProperty {α : Type _} [DecidableEq α] (Sys : Finset (Finset α)) :=
+  {s₁ : Finset α} → s₁ ∈ Sys →
+  {s₂ : Finset α} → s₂ ∈ Sys →
+  s₂.card < s₁.card →
+  ∃ x ∈ s₁ \ s₂, insert x s₂ ∈ Sys
```

Comment:
I'm sorry that your PR sits here for a long time with no reviews.
It seems that most (all?) of Mathlib maintainers don't feel to be experts in this area...

I know nothing about greedoids, so I'm going to leave some purely technical remarks.

In this definition, usage of `{}` binders may lead to surprising results like "Lean eats an argument when you don't expect it". This was definitely the case with Lean 3 and may be the case with Lean 4. I suggest using `⦃⦄` to prevent this.

Also, it's possible to formulate this definition without a `DecidableEq` assumption.

```suggestion
def ExchangeProperty {α : Type*} (Sys : Finset (Finset α)) : Prop :=
  ∀ ⦃s : Finset α⦄, s ∈ Sys → ∀ ⦃t : Finset α⦄, t ∈ Sys → t.card < s.card →
    ∃ x ∈ s, ∃ h : x ∉ t, cons x t h ∈ Sys
```

Reply:
Thanks for the reply!
Frankly speaking, I'm also not an expert in this field. I am also reading [this book](https://link.springer.com/book/10.1007/978-3-642-58191-5) at my own pace to study more about greedy algorithms (although this led me to a whole new world). So the upcoming definitions and theorems will also follow the order of the book. In other words, I'm very happy to get any technical reviews, thanks a lot!